compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -1703,13 +1704,32 @@ fn add_post_link_args(cmd: &mut dyn Linker, sess: &Session, flavor: LinkerFlavor
 /// that are necessary for the linking. They are only present in symbol table but not actually
 /// used in any sections, so the linker will therefore pick relevant rlibs for linking, but
 /// unused `#[no_mangle]` or `#[used]` can still be discard by GC sections.
+///
+/// There's a few internal crates in the standard library (aka libcore and
+/// libstd) which actually have a circular dependence upon one another. This
+/// currently arises through "weak lang items" where libcore requires things
+/// like `rust_begin_unwind` but libstd ends up defining it. To get this
+/// circular dependence to work correctly we declare some of this things
+/// in `symbols.o` file.
 fn add_linked_symbol_object(
     cmd: &mut dyn Linker,
     sess: &Session,
+    info: &CrateInfo,
     tmpdir: &Path,
     symbols: &[(String, SymbolExportKind)],
 ) {
-    if symbols.is_empty() {
+    let weak_symbols = (&info.used_crates)
+        .into_iter()
+        .filter_map(|cnum| info.missing_lang_items.get(&cnum))
```

Comment:
I think `missing_lang_items` can now be made a single set as opposed to a map from crate id to a list of missing lang items. I also think `lang_item_to_crate` is no longer necessary as of this PR.

Reply:
Yeah, @BelovDV, please consider reimplementing these tables so that they contain only what we actually need now, instead of reusing `missing_lang_items`/`lang_item_to_crate` as they exist.

In the end we need a function from "weak lang item symbol name" to "whether it's provided by any non-`Linkage::NotLinked` dependency and required by any non-`Linkage::NotLinked` dependency".

(I still don't understand what a "missing lang item" means, it seems to only cover the "required" part, but not "provided". I'll look at the existing code in more detail a bit later.)

Reply:
"Missing lang item" for a crate means that the weak lang item was missing in the respective crate and all of it's dependencies when compiling it, but would need to be provided by another crate.

Reply:
I missed one important thing that makes everything much simpler - `weak_lang_items::verify` reports an error in any weak lang item is missing, for all crate types except rlibs.
So if we have any linked crate type then all missing items are also provided by definition, so we can indeed just build a union of all `missing_lang_items` from all crates and add it to symbols.o.